### PR TITLE
remove MysqlResultSet.Name2Index

### PR DESCRIPTION
**Pull request by CJKkkk-315**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [x] Code Refactoring

## Which issue(s) this PR fixes:

issue #12390 

## What this PR does / why we need it:

Removed Name2Idex from MysqlResultSet
Now, every operation related to Add or Remove Columns in MysqlResultSet requires corresponding map operation in Name2Idex. 
Additionally, creating MysqlResultSet and copying both require additional memory overhead to make (map).
But there are almost no scenarios where Name2Idex is necessary, removing it can optimize mem and CPU overhead.

### CODE REVIEW

File: pkg/stream/connector/converter.go [hunk 1]
```diff
@@ -62,19 +60,18 @@ func (c *SQLConverter) Convert(ctx context.Context, obj ie.InternalExecResult) (
...
-		colNames = append(colNames, name)
 	}
 	for i := 0; i < rowCount; i++ {
 		var rowValues string
 		var err error
 		for j := 0; j < columnCount; j++ {
-			var val string
-			val, err = obj.StringValueByName(ctx, uint64(i), colNames[j])
+			var val interface{}
+			val, err = obj.Value(ctx, uint64(i), uint64(j))
```
Comment (daviszhen) [inline pkg/stream/connector/converter.go:69]: 这地方用GetString把

File: pkg/util/metric/mometric/cron_task.go [hunk 2]
```diff
@@ -203,19 +206,45 @@ func CalculateStorageUsage(
...
+				}
+				name2idx[colName] = colIdx
+			}
+		}
+
 		for rowIdx := uint64(0); rowIdx < cnt; rowIdx++ {
-			account, err := result.StringValueByName(ctx, rowIdx, ColumnAccountName)
-			if err != nil {
-				return err
+			if accountIdx, ok := name2idx[ColumnAccountName]; !ok {
```
Comment (daviszhen) [inline pkg/util/metric/mometric/cron_task.go:221]: 如果show accounts 的结果列是固定的。accound，size,snapshotSize的位置是固定的。不用再建立map，每次查询了。直接定义成常量就可以了吧。

File: pkg/util/metric/mometric/cron_task.go [hunk 3]
```diff
@@ -302,16 +334,36 @@ func checkNewAccountSize(ctx context.Context, logger *log.MOLogger, sqlExecutor
 			goto nextL
 		}
 		logger.Debug("collect new account cnt", zap.Uint64("cnt", newAccountCnt))
+
+		if name2idx == nil {
+			name2idx = make(map[string]uint64)
```
Comment (daviszhen) [inline pkg/util/metric/mometric/cron_task.go:339]: 原因与上面类似。

**Review by daviszhen — CHANGES_REQUESTED** [review]

File: pkg/util/metric/mometric/cron_task.go (hunk 2 above)
Comment (xzxiong) [inline pkg/util/metric/mometric/cron_task.go:221]: 代码逻辑需要调整。
1. 需要在 L209 行，完成对 accountIdx 和 sizeIdx 和 snapshotIdx 的计算，且只需要计算一次。
2. L221 行，直接取值，不判断是否存在，不允许默认值。（除了 snapshot_size字段）
3. 目前只有 snapshotSizeIdx 允许为 -1, 其他都必须存在。否则err return

前提条件：
1. account，size，snapshot_size 都是必须有的。
4. 可选的snapshot_size 只是版本过渡，便于mocloud 测试。

File: pkg/util/metric/mometric/cron_task.go [hunk 4]
```diff
@@ -203,19 +206,45 @@ func CalculateStorageUsage(
 		}
 		logger.Debug("collect storage_usage cnt", zap.Uint64("cnt", cnt))
 		metric.StorageUsageFactory.Reset()
```
Comment (xzxiong) [inline pkg/util/metric/mometric/cron_task.go:208]: 这一行需要挪到 name2idx 下边。

File: pkg/util/metric/mometric/cron_task.go [hunk 5]
```diff
@@ -302,16 +334,36 @@ func checkNewAccountSize(ctx context.Context, logger *log.MOLogger, sqlExecutor
 			goto nextL
 		}
 		logger.Debug("collect new account cnt", zap.Uint64("cnt", newAccountCnt))
+
+		if name2idx == nil {
```
Comment (xzxiong) [inline pkg/util/metric/mometric/cron_task.go:338]: 同上，accountIdx, sizeIdx, snapshotSizeIdx 仅需计算一次。

File: pkg/util/metric/mometric/cron_task.go (hunk 4 above)
Comment (xzxiong) [reply pkg/util/metric/mometric/cron_task.go:249]: ref: https://github.com/CJKkkk-315/matrixone/pull/3

File: pkg/util/metric/mometric/cron_task.go (hunk 5 above)
Comment (xzxiong) [reply pkg/util/metric/mometric/cron_task.go]: fixed by https://github.com/CJKkkk-315/matrixone/pull/3

File: pkg/util/metric/mometric/cron_task.go (hunk 2 above)
Comment (xzxiong) [reply pkg/util/metric/mometric/cron_task.go]: fixed by https://github.com/CJKkkk-315/matrixone/pull/3

**Review by daviszhen — APPROVED** [review]

**Review by xzxiong — APPROVED** [review]

**Review by aptend — APPROVED** [review]

**Review by fengttt — CHANGES_REQUESTED** [review]
I don't have any clue what is this, why we need it, and what benefit we will see from this.

Need a better commit message.   Explaining.    If any "real" benefit, there should be test to show off it.

**Comment by CJKkkk-315** [comment]
> I don't have any clue what is this, why we need it, and what benefit we will see from this.
> 
> Need a better commit message. Explaining. If any "real" benefit, there should be test to show off it.

A more detailed explanation has been provided in "What this PR does / why we need it"

**Review by fengttt — APPROVED** [review]

**Pull request merged**